sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/HCatRecordReaderFn.java

```diff
@@ -0,0 +1,165 @@
+    for (int i = 0; i < partitionCols.size(); i++) {
+      filter.add(partitionCols.get(i) + "=" + "'" + values.get(i) + "'");
+    }
+    final String filterString = String.join(" and ", filter);
+
+    ReadEntity entity =
+        new ReadEntity.Builder()
+            .withDatabase(readRequest.getDatabase())
+            .withFilter(filterString)
+            .withTable(readRequest.getTable())
+            .build();
+    // pass the 'desired' split count as an hint to the API
+    Map<String, String> configProps = new HashMap<>(readRequest.getConfigProperties());
+    configProps.put(
+        HCatConstants.HCAT_DESIRED_PARTITION_NUM_SPLITS, String.valueOf(desiredSplitCount));
+    return DataTransferFactory.getHCatReader(entity, configProps).prepareRead();
+  }
+
+  private long getFileSizeForPartitions(Read readRequest) throws Exception {
+    IMetaStoreClient client = null;
+    try {
+      HiveConf hiveConf = HCatalogUtils.createHiveConf(readRequest);
+      client = HCatalogUtils.createMetaStoreClient(hiveConf);
+      List<org.apache.hadoop.hive.ql.metadata.Partition> p = new ArrayList<>();
+      Table table = HCatUtil.getTable(client, readRequest.getDatabase(), readRequest.getTable());
+      final org.apache.hadoop.hive.ql.metadata.Partition partition =
+          new org.apache.hadoop.hive.ql.metadata.Partition(table, readRequest.getPartitionToRead());
+      p.add(partition);
+      final List<Long> fileSizeForPartitions = StatsUtils.getFileSizeForPartitions(hiveConf, p);
+      return fileSizeForPartitions.get(0);
+    } finally {
+      // IMetaStoreClient is not AutoCloseable, closing it manually
+      if (client != null) {
+        client.close();
+      }
+    }
+  }
+
+  /** Reads data for a specific partition. */
+  @ProcessElement
+  @SuppressWarnings("unused")
+  public void processElement(
+      ProcessContext processContext, RestrictionTracker<OffsetRange, Long> tracker)
+      throws Exception {
+    final Read readRequest = processContext.element();
+    final List<String> values = readRequest.getPartitionToRead().getValues();
+    final String watermarkPartitionColumn = readRequest.getWatermarkPartitionColumn();
+    final int indexOfPartitionColumnWithWaterMark =
+        watermarkPartitionColumn.indexOf(watermarkPartitionColumn);
+    final String partitionWaterMark = values.get(indexOfPartitionColumnWithWaterMark);
+
+    final Instant partitionWatermarkTimeStamp =
+        readRequest.getWatermarkTimestampConverter().apply(partitionWaterMark);
+
+    int desiredSplitCount = 1;
+    long estimatedSizeBytes = getFileSizeForPartitions(processContext.element());
+    if (estimatedSizeBytes > 0) {
+      desiredSplitCount = (int) Math.ceil((double) estimatedSizeBytes / Integer.MAX_VALUE);
+    }
+    ReaderContext readerContext = getReaderContext(processContext.element(), desiredSplitCount);
+
+    for (long i = tracker.currentRestriction().getFrom(); tracker.tryClaim(i); i++) {
```

Comment:
Hmm why does this one need to be an SDF rather than simply a composition of two DoFn's with a reshuffle in between? (where the first DoFn would emit (Read, int) pairs where the int is 0..numSplits) Is there so many splits that it is prohibitive?